Clarify note about eager vs. lazy evaluation (#901)
unicode-org/message-format-wg: spec/formatting.md

> _declarations_ affecting _variables_ referenced by that _expression_
> have already been evaluated in the order in which the relevant _declarations_
> appear in the _message_.
>

> > Users may write custom functions that have observable side effects.
> > Lazy evaluation may involve evaluating the same expression multiple times

Reviewer: It's good practice to avoid the normative words like 'may' in non-normative contexts like this.

```suggestion
> Users or implementations can provide functions that have observable side effects.
> Lazy evaluation might involve evaluating the same _expression_ multiple times
```

Notice that I removed the word "custom" (which might be presumed?) and bring in implementations. 

Consider adding a health warning to this paragraph?

Author: Removed the "may". What do you mean by "health warning"?

Reviewer: By health warning, something like: It's a bad idea to create selectors or formatter functions with such side effects

Author: I see. I'm not sure I would agree across the board. For example, you might want a `currentTime` function that returns the system clock time. This doesn't have "side effects" as such, but returns a different value when it's called at different times, so:

```
.local $time = {:currentTime}
{{ {$time} {$time} }}
```

could have a different result depending on whether the implementation evaluates the RHS of the variable `time` once or twice.

(The jargon is that `currentTime` is not referentially transparent.)

Perhaps some language saying that care should be taken when defining such functions and they should be avoided when possible?

Reviewer 2: I thought we already had something in the spec about side effects, but the closest we come to is
> _Function handler_ access to the _formatting context_ MUST be minimal and read-only,
and execution time SHOULD be limited.

which relies on an implementation choosing to call "everything that a function can access" as its _formatting context_, because we've not explicitly prohibited access outside it.

That is to say, I agree with @aphillips that we ought to have stronger language to recommend against functions that mutate anything as a side effect.

Author: I added some more language to (a) clarify that the problem also occurs with functions that don't mutate anything, but that depend on external mutable state; and (b) recommend against functions that mutate anything observably.

Reviewer: Hmm...

side effect == touches the _resolved value_?

Lots of functions have "side effects" if that's the case. @eemeli your suggestion of limiting the _resolved value_ of `:date` and `:time` would be a side effect if we were to implement it. `currentTime` is a good example. `random` is another. Some operations, such as case-folding, alter the value in an irreversible way. Other operations, such as unfloating a time value, change the meaning.

Our design tries to be immutable, but we're actually mutable to the depth of one expression. Chaining `.local` or reannotating a placeholder can make these side-effects more visible.

Perhaps this paragraphs wants to be more like:

> Evaluating an _expression_ can affect the _resolved value_ associated with 
> a given _variable_.
> When combined with eager vs. lazy evaluation, 
> in which the same _expression_ might be evaluated multiple times,
> the side-effects of the _function_ on the _resolved value_
> might alter the result of formatting the _message_.

Reviewer 2: > side effect == touches the _resolved value_?
>
> Lots of functions have "side effects" if that's the case. @eemeli your suggestion of limiting the _resolved value_ of `:date` and `:time` would be a side effect if we were to implement it.

Defining the _resolved value_ returned by a _function handler_ is not a [side effect](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). Modifying the _operand_ would be a side effect. But a _function handler_ will quite rarely have a reason to return its input _operand_; in general, it will instead create a new _resolved value_ and return that.

> `currentTime` is a good example. `random` is another. Some operations, such as case-folding, alter the value in an irreversible way. Other operations, such as unfloating a time value, change the meaning.

A reasonable _function handler_ for any of these would not have any side effects, but would instead create a new _resolved value_ and return it.

> Our design tries to be immutable, but we're actually mutable to the depth of one expression. Chaining `.local` or reannotating a placeholder can make these side-effects more visible.

Being able to use a _resolved value_ as an _operand_ does not introduce mutability.

Author: The phrase "side effect" is clearly confusing, so I rewrote the paragraph without using it. Please take another look!

> _declarations_ affecting _variables_ referenced by that _expression_
> have already been evaluated in the order in which the relevant _declarations_
> appear in the _message_.
>

> > Implementations and users creating custom functions SHOULD avoid
> > creating _function handlers_ that mutate external program state.

Reviewer: SHOULD avoid => SHOULD NOT

```suggestion
> Implementations and users SHOULD NOT create _function handlers_
> that mutate external program state,
> particularly since such a _function handler_ can present a remote execution hazard.
```

>

> [!NOTE]
>
> Function handlers can also be written that depend on external mutable state
> (for example, the current date and time).
> Lazy evaluation might involve evaluating the same _expression_ multiple times

> > (call-by-name) or evaluating every expression at most once (call-by-need).

Reviewer 3: I really think it's a mistake to allow the same expression to ever evaluate to a different value. This is something that [CEL](https://github.com/google/cel-spec/blob/master/doc/langdef.md#evaluation) (to pick a roughly analogous technology) gets _very_ right.

> _declarations_ affecting _variables_ referenced by that _expression_
> have already been evaluated in the order in which the relevant _declarations_
> appear in the _message_.
> An implementation MUST ensure that every _expression_ in a _message_
> is evaluated at most once.

> >
> > Implementations and users SHOULD NOT create _function handlers_

Reviewer: This is a separate consideration. It should be a separate note.

```suggestion

> [!IMPORTANT]
> Implementations and users SHOULD NOT create _function handlers_
```

> that mutate external program state,
> particularly since such a _function handler_ can present a remote execution hazard.
>

> [!NOTE]
>
> Function handlers can be written that depend on external mutable state
> (for example, the current date and time).
> Lazy evaluation might involve evaluating the same _expression_ multiple times
> (call-by-name) or evaluating every expression at most once (call-by-need).
> This justifies the requirement to ensure that every _expression_ is
> evaluated at most once: different invocations of the same function handler
> might return different results (functions are not pure).
> This means that lazy implementations cannot use

> > a call-by-name evaluation strategy.

Reviewer: I think this note is burying the lede. It should appear directly after the (first) "IMPORTANT", where we forbid multiple evaluations and I think it should start with the guidance. Something like:

> [!NOTE]
> Implementations that choose lazy evaluation cannot use a call-by-name
> evaluation strategy (instead of evaluating expressions at most once, or 
> "call-by-need").
> This is to prevent _function handlers_ that depend on external mutable state
> (for example, ones that obtain the current system clock time)
> from returning different values for the same _expression_ in different
> parts of a given _message_.

Author: Took another pass at it in [fdbb48c](https://github.com/unicode-org/message-format-wg/pull/901/commits/fdbb48c57ca804e652902bb05951a98f64d29f6f) -- let me know what you think.

> have already been evaluated in the order in which the relevant _declarations_
> appear in the _message_.
>
> In some environments, users or implementations can provide functions
> that mutate state that is external to the message formatter

> > (for example, deleting a file in the filesystem). This is not recommended.

Reviewer: "not recommended" turns out to be a 2119 keyword 😉 

Is this a likely example? Maybe a better example would be modifying the system or application locale (in some environments, setting the locale is not thread-safe)

```suggestion
(for example, setting or modifying the system locale).
```

Reviewer: Actually... I think the problem here is the audience for your addition. This turns out not to be in a note block. It's in a normative "IMPORTANT" section. It should be written in a normative way or it should be in an informative NOTE inside the IMPORTANT block.

I'd actually do the former, replacing this paragraph with something like:

> Implementations and users creating custom functions SHOULD avoid 
> creating _function handlers_ that mutate external program state
> or which depend on external mutable state in a way that would
> result in different _resolved values_ for a _variable_ depending on 
> whether a given _expression_ is evaluated at most once (call-by-need)
> or multiple times (call-by-name).

The `:currentDateTime` is a good example of a function that implementation might very well want to provide and which produces a different result every time you call it (but at least a few ms, right?). Users of that function would need to know that they can't rely on it producing the exact same number each time they call it (the system clock is running) or that they can't rely on it to _not_ provide the same value every time in a given formatting call (the system clock is running, but the function only evaluates it once).

The warning about mutating program state doesn't seem to belong with eager/lazy? Can you think of an example of a real function that an implementation might provide (which isn't a security exploit)? I mean, my "set locale" suggestion above might quality in C, for example. That would certainly affect the formatting of the message (unless the "formatting context" was caching the locale). It would also be cause for spanking the developer who created/installed it.

Author: I follow, but I'm a little unsure as to what you're suggesting the normative change should be. If it's "SHOULD avoid...", as in your suggested wording, doesn't that suggest there shouldn't be a `currentDateTime` or `random` function? But I can imagine these functions being useful, it's just that it may be surprising that different implementations with different evaluation strategies would handle them differently. Do we want to say to not provide functions like these, or just to be judicious about it?

Author: I tried splitting the note into a normative part and an editorial "note" part, maybe that's closer to what you have in mind?
